### DISCUSSION TITLE
Are there tools similar to the ast visitor pattern？

### DISCUSSION THREAD

**Discussion by xin2017338** [oxc-project/oxc#1177] [Q&A] [answered]
Hey, when I was looking at oxc code and writing lint, I realized why there wasn't a design like eslint visitor pattern. For no-this-before-super, it was tedious to implement that I needed to store the call stack.

**Comment by Boshen** [comment, marked as answer]
no-this-before-super requires a control flow graph, which we don't have yet.

**Reply by rizrmd** [reply]
oxc-parser is really fast, but still tedious to use. A visitor function would really help us to traverse the ast.

**Comment by Boshen** [comment]
https://github.com/oxc-project/oxc/blob/main/crates/oxc_linter/src/rules/eslint/no_this_before_super.rs
